Add summaries for the security models of attestations and trusted publishers (#17242)
pypi/warehouse: docs/user/trusted-publishers/security-model.md

# Security model and considerations

> Trusted publishing is primarily designed to be a more secure alternative than

Reviewer: Nitpick: we generally capitalize it as a term of art:

```suggestion
Trusted Publishing is primarily designed to be a more secure alternative to
```

the long-lived API tokens that have traditionally been used for publishing to
PyPI.

In recent years, theft of credentials such as API tokens has [played a major
role in cyber attacks]. The reason for this is the unfortunate reality that

> managing credentials can be unexpectedly difficult. Trusted publishing addresses

Reviewer: ```suggestion
managing credentials can be unexpectedly difficult. Trusted Publishing addresses
```

this risk by using short-lived tokens instead of long-lived tokens that can
easily get misplaced, leaked in logs, stolen by malware, or any number of other
ways.

However, it is important to still be aware of the kinds of risks that

> trusted publishing does not cover. You should think of trusted publishing as one

Reviewer: ```suggestion
Trusted Publishing does not cover. You should think of Trusted Publishing as one
```

tool in the toolbelt for securing packages.

## General considerations

> * Trusted publishing asserts that a package was published to PyPI by someone
>   or something that had access to an API token that was created within 15
>   minutes of a valid OIDC authentication flow. It does not assert anything about
>   the safety of the code or the trustworthiness of the authors.

Reviewer: The wording here is a little confusing IMO. Suggestion:

```suggestion
* Trusted Publishing uses short-lived API tokens that expire 
  no more than 15 minutes after the OIDC flow that authorizes them.
  Just like normal API authentication, Trusted Publishing
  does not assert the safety of the code or the trustworthiness
  of its authors.
```

> * Trusted publishing does not address whether the package has been modified

Reviewer: ```suggestion
* Trusted Publishing does not address whether the package has been modified
```

  before or after it was built. [Attestations] can address those risks.

* Short-lived API tokens are sensitive material that must be protected from
  getting stolen or leaked.

* OIDC tokens themselves are also sensitive material that must be protected
  from getting stolen or leaked. OIDC tokens expire quickly, but an attacker who
  successfully intercepts one can use it to generate API tokens until it
  expires.

> * Configuring a trusted publisher means trusting in the auth system of a
>   particular OIDC provider (such as GitHub Actions). Trusted publishing relies
>   on that auth system not being contolled by attackers and not containing
>   vulnerabilities.

Reviewer: ```suggestion
* Configuring a Trusted Publisher means trusting an identity provider (IdP),
  such as GitHub Actions. Trusted Publishing relies on the integrity of that
  IdP and the actors that are authorized to use it. In practice, this means
  that users of Trusted Publishing must protect and secure the CI/CD workflows
  that they register as Trusted Publishers, as weaknesses in those workflows
  can be equivalent to credential compromise.
```

> In summary: treat your trusted publishers *as if* they were API tokens. If you

Reviewer: (This one is my fault from an earlier version of the doc!)

```suggestion
In summary: treat your Trusted Publishers *as if* they are API tokens. If you
```

# Security model and considerations

Trusted Publishing is primarily designed to be a more secure alternative to
the long-lived API tokens that have traditionally been used for publishing to
PyPI.

In recent years, theft of credentials such as API tokens has [played a major
role in cyber attacks]. The reason for this is the unfortunate reality that
managing credentials can be unexpectedly difficult. Trusted Publishing addresses
this risk by using short-lived tokens instead of long-lived tokens that can
easily get misplaced, leaked in logs, stolen by malware, or any number of other

> ways.

Reviewer 2: I think the key distinction we should make here is that long-lived tokens need to be stored and do not quickly expire. That somewhat reduces the likelihood that they will "misplaced, leaked in logs, stolen by malware" (although it's still possible) but _also_ reduces the impact if they are compromised.

Author: Good point, let me know what you think of the new version I just pushed 👍

# Security model and considerations

Trusted Publishing is primarily designed to be a more secure alternative to
the long-lived API tokens that have traditionally been used for publishing to
PyPI.

In recent years, theft of credentials such as API tokens has [played a major
role in cyber attacks]. The reason for this is the unfortunate reality that
managing credentials can be complicated and risky. Trusted Publishing reduces
this risk by using short-lived tokens instead of long-lived tokens. Short-lived
tokens are less likely to be misplaced, leaked in logs, or stolen by malware.
If short-lived tokens are leaked, they only give attackers a narrow time window

> to exploit the leaked token, which minimizes the potential damage.

Reviewer 2: I think this is still missing the distinction that when used with Trusted Publishing short-lived tokens are ephemeral and don't need to be stored.

Author: Oh I thought you wanted emphasis on the fact that short-lived tokens only reduce but don't eliminate the risk. How about this?

> Short-lived tokens are less likely to be misplaced, leaked in logs, or stolen by malware since they don't have to be stored.

Reviewer 2: Perfect!

> tokens are less likely to be misplaced, leaked in logs, or stolen by malware.
> If short-lived tokens are leaked, they only give attackers a narrow time window
> to exploit the leaked token, which minimizes the potential damage.

Reviewer 2: ```suggestion
tokens are less likely to be misplaced, leaked in logs, or stolen by malware
since they don't have to be stored. Additionally, if short-lived tokens are
leaked, they only give attackers a narrow time window to exploit the leaked
token, which minimizes the potential damage.
```

      access the OIDC token to a bare minimum. This prevents both accidental
      and malicious disclosure.

> [played a major role in cyber attacks]: https://verizon.com/dbir

Reviewer: This HTTPS link doesn't connect successfully for me, but the HTTP version redirects to <https://www.verizon.com/business/resources/reports/dbir/>. Perhaps we should update it to that?

Reviewer: Looks like the actual report is [here](https://www.verizon.com/business/resources/reports/2024-dbir-data-breach-investigations-report.pdf), but it's emailwalled. I think we should probably find an alternative source/cite, since PyPI's docs shouldn't link people to walled sources 🙂 

Here are some candidates:

* <https://therecord.media/cisa-cyberattacks-using-valid-credentials>
* <https://www.cisa.gov/sites/default/files/2023-07/FY22-RVA-Analysis%20-%20Final_508c.pdf>

Author: The email-walled version is unfortunately the one the authors ask that we link to. But that first link looks good, will replace it.
